### PR TITLE
Fixes and improvements for numbered problems

### PR DESCRIPTION
**Pull request by bjones1**
Fix: Restart numbering at 1 on every page.
Fix: Directives that begin with a directive now render correctly.
Add: Visually-checked example of a directive which begins with a directive.

### CODE REVIEW

**Comment by bnmnetp** [comment]
Are we sure this is what we want to do?  I could see a use case where people would want to number exercises sequentially for an entire Chapter rather than starting at 1 on every subchapter.  I think this depends a little on whether you prefer to write with long pages or lots of short pages.  But I could see it getting confusing if there were multiple list-q-1 in the lists chapter.

**Comment by bjones1** [comment]
The current system is broken; this at least gives consistent numbering. In particular, consider this case:

```
Chapter 1
  Subchapter A
  Subchapter B
  Subchapter C
```

After a `runestone build --all`, everything is OK. Then, an edit of subchapter B will produce renumbered problems in that subchapter only -- and the renumbering will either start at 1 (if the previous numbering wasn't saved in `env`) or the next number after the last problem on subchapter C (if not -- I don't remember which). Adding or removing problems to subchapter B should cause a renumbering of subchapter C, but it doesn't. Fixing these problems seem fairly challenging to me, so I opted for the simpler fix presented here.

File: runestone/common/runestonedirective.py [hunk 1]
```diff
@@ -94,6 +98,10 @@ def _purge_runestone_data(app, env, docname):
         runestone_data.id_to_page.pop(id_)
     runestone_data.page_to_id.pop(docname)
 
+    # Reset the problem numbering for each page.
+    env.assesscounter = 0
```
Comment (bnmnetp) [inline runestone/common/runestonedirective.py]: if the .. qnum directive explicitly sets the starting value for this chapter will this override it??

**Comment by bnmnetp** [comment]
OK, good point.  We actually addressed that in fopp by setting the prefix for each subchapter.

I just want to make sure this doesn't override an explicitly set starting number for a chapter.

File: runestone/common/runestonedirective.py (hunk 1 above)
Comment (bjones1) [reply runestone/common/runestonedirective.py]: No, this occurs before any directives on the page are processed. It's used to purge any stale data for the page (hence the name).

**Comment by bjones1** [comment]
Sounds like a good chance to test the `qnum` directive. I can add this to the unit test (although checking it will be harder...)

**Comment by bjones1** [comment]
I've just pushed updates which includes tests of `qnum`, plus does better checking of the directive-in-directive case. This should be ready to go!

**Comment by bnmnetp** [comment]
Thanks!

**Pull request merged**